File: 数据结构课设/DataStructure/LintCode_easy/LinearList/insertionSortList.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""


class Solution:
    """
    @param: head: The first node of linked list.
    @return: The head of linked list.
    """
    def insertionSortList(self, head):
        # write your code here
        if head == None:
            return head
        #first 为未排序链表的头节点
        first = head.next
        head.next = None
        #设置一个dummy链接在head的前面
        dummy = ListNode(0)
        dummy.next = head
        #设置一个尾节点和新值比较
        tail = head
        while first:
            #新值比尾节点要大则直接接上
            if tail.val < first.val:
                tmp = first
                tail.next = tmp
                first = first.next
                tail = tail.next
                tail.next = None
            #新值要是比尾节点小则从头开始遍历
            else:
                pre = dummy
                cur = dummy.next
                while cur:
                    if cur.val < first.next:
                        pre = pre.next
                        cur = cur.next
                    else:
                        tmp = first
                        first = first.next
                        tmp.next = cur
                        pre.next = tmp
                        break
        return dummy.next

File: 数据结构课设/DataStructure/LintCode_easy/LinearList/insertionSortList.cpp
/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */


class Solution {
public:
    /*
     * @param head: The first node of linked list.
     * @return: The head of linked list.
     */
    ListNode * insertionSortList(ListNode * head) {
        // write your code here
        if(head == NULL){
            return head;
        }
        ListNode *newHead = new ListNode(INT_MIN);
        newHead->next = head;
        ListNode *alSort, *noSort, *p, *q;
        alSort = newHead->next;
        newHead->next = NULL;
        while(alSort != NULL){
            p = newHead;
            noSort = newHead->next;
             while (noSort != NULL && noSort->val <= alSort->val) {
                p = noSort;
                noSort = noSort->next;
            }
            q = alSort->next;
            alSort->next = p->next;
            p->next = alSort;
            alSort = q;
        }
        return newHead->next;
        }
    
};
